#!/bin/bash

#
# Post-processing for vcf files generated by bcftools calling pipeline: clean-up and filtering
#


#f=$1
f=GR_freeze2019.vcf.gz
echo "Post-processing $f"

# clean-up the header
bcftools view -h ${f} | grep -v "HLA" | egrep -v "^##contig.*chr.*_"  | grep -v "##bcftools_concatCommand" | grep -v "chrEBV" > ${f}.new_header.txt

for chr in `seq 1 22` X Y MT
do
 echo "chr=$chr"

 # Filter QUAL > 40, GQ > 20, DP > 10, SP < 20
 # Remove sites with low mappability (downloaded from UMAP website)
 # remove chr prefix from chromosome names
 # requires bioformats (by Gaik Tamazian)
 bioformats filter_variants_vcf 40 20 10 20 <( bcftools view -R mappability_per_chr/k100.umap.chr_prefix.chr${chr}.no_overlap.bed.gz ${f} ) | bcftools reheader -h ${f}.new_header.txt  | bcftools annotate --rename-chrs /mnt/gr1/resources/chr_renaming.txt -Oz > ${f%vcf.gz}chr${chr}.QUAL40_GQ20_DP10_SP20.vcf.gz

 tabix -p vcf ${f%vcf.gz}chr${chr}.QUAL40_GQ20_DP10_SP20.vcf.gz

 # Add rs ids from dbSNP
 echo "add rs ids"
 bcftools annotate -a /mnt/gr1/resources/All_20180418.vcf.gz -c ID -Oz ${f%vcf.gz}chr${chr}.QUAL40_GQ20_DP10_SP20.vcf.gz > ${f%vcf.gz}chr${chr}.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz
 tabix -p vcf ${f%vcf.gz}chr${chr}.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz

done

# concatenate per chromosome files
bcftools concat -Oz -o ${f%vcf.gz}all_chr.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz GR_freeze2019.chr*.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr2.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr3.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr4.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr5.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr6.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr7.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr8.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr9.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr10.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr11.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr12.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr13.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr14.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr15.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr16.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr17.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr18.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr19.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr20.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr21.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chr22.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chrX.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz ${f%vcf.gz}chrY.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz

# exclude 2 outliers (high missingness rate) and keep only sites with at least one alternative allele
bcftools view -s ^GR1756,GR0814 --min-ac 1 -Oz ${f%vcf.gz}all_chr.QUAL40_GQ20_DP10_SP20.rsids.vcf.gz > ${f%vcf.gz}all_chr.QUAL40_GQ20_DP10_SP20.rsids.final.min-ac1.vcf.gz 
tabix -p vcf ${f%vcf.gz}all_chr.QUAL40_GQ20_DP10_SP20.rsids.final.min-ac1.vcf.gz 

ln -s ${f%vcf.gz}all_chr.QUAL40_GQ20_DP10_SP20.rsids.final.min-ac1.vcf.gz GR_freeze2019.filtered.vcf.gz
ln -s ${f%vcf.gz}all_chr.QUAL40_GQ20_DP10_SP20.rsids.final.min-ac1.vcf.gz.tbi GR_freeze2019.filtered.vcf.gz.tbi

#
# convert to Plink
#

# Add family ids, split the parX regions
plink --vcf GR_freeze2019.filtered.vcf.gz --split-x b38 --make-bed --out ${out_base}.splitx --update-ids /mnt/sochi/dzhernakova/GR/GR_freeze2019/sample_info/update_fid_iid.txt
# Add mother and father ids
plink --bfile ${out_base}.splitx --update-parents /mnt/sochi/dzhernakova/GR/GR_freeze2019/sample_info/update_parents.txt --update-sex /mnt/sochi/dzhernakova/GR/GR_freeze2019/sample_info/update_sex.txt --make-bed --out ${out_base}.splitx.pheno

echo "Finished!"